### PR TITLE
Skip negotiation

### PR DESCRIPTION
**Pull request by wegylexy**
See https://github.com/dotnet/aspnetcore/issues/21386

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Thanks for the PR, sorry we haven't taken a look yet.

It looks like there might be a hang somewhere or the CI conveniently failed on both Windows runs. I'm rerunning the check to see if it's an actual hang.

**Comment by wegylexy** [comment]
It also says `process took too long` for Mac and Linux, but the test ultimately passed after that timeout of `url.negotiate_appended_to_url`.

**Comment by BrennanConroy** [comment]
Ok, looks like there is a hang then. Also a bug in our timeout logic that should have failed the build step instead of passing 😆

**Comment by wegylexy** [comment]
Any clue?

**Comment by BrennanConroy** [comment]
Well I'm assuming the hang is caused by the PR since it doesn't happen on other builds. So I haven't taken a look.

**Comment by BrennanConroy** [comment]
Starting to look at this again, want to see what the current CI situation is with the changes.

File: include/signalrclient/hub_connection_builder.h [hunk 1]
```diff
@@ -42,5 +44,6 @@ namespace signalr
         trace_level m_log_level;
         std::function<std::shared_ptr<websocket_client>(const signalr_client_config&)> m_websocket_factory;
         std::shared_ptr<http_client> m_http_client;
+        bool m_skip_negotiation;
```
Comment (BrennanConroy) [inline include/signalrclient/hub_connection_builder.h:47]: This is currently uninitialized which is causing errors
```suggestion
        bool m_skip_negotiation = false;
```

File: src/signalrclient/connection_impl.cpp [hunk 2]
```diff
@@ -155,8 +155,64 @@ namespace signalr
...
+
+                assert(false);
+            }
+
+            connection->m_start_completed_event.cancel();
+            callback(nullptr);
+        };
+
+        if (m_skip_negotiation)
+            return start_transport(url, started);
```
Comment (BrennanConroy) [inline src/signalrclient/connection_impl.cpp:212]:
```suggestion
        if (m_skip_negotiation)
        {
            return start_transport(url, started);
        }
```

File: src/signalrclient/connection_impl.cpp [hunk 3]
```diff
@@ -155,8 +155,64 @@ namespace signalr
         std::weak_ptr<connection_impl> weak_connection = shared_from_this();
         const auto& token = m_disconnect_cts;
 
+		const auto started = [weak_connection, callback, token](std::shared_ptr<transport> transport, std::exception_ptr exception)
```
Comment (BrennanConroy) [inline src/signalrclient/connection_impl.cpp:158]: I don't see why this was moved, could we revert it if it's unrelated to the change.

Comment (wegylexy) [reply src/signalrclient/connection_impl.cpp:158]: To be captured in the lambda.

Comment (BrennanConroy) [reply src/signalrclient/connection_impl.cpp:158]: Oh I see, I missed that it was being used in two places.
```suggestion
        const auto transport_started = [weak_connection, callback, token](std::shared_ptr<transport> transport, std::exception_ptr exception)
```

File: src/signalrclient/connection_impl.cpp [hunk 4]
```diff
@@ -155,8 +155,64 @@ namespace signalr
...
+            connection->m_start_completed_event.cancel();
+            callback(nullptr);
+        };
+
+        if (m_skip_negotiation)
+            return start_transport(url, started);
+
         negotiate::negotiate(*m_http_client, url, m_signalr_client_config,
-            [callback, weak_connection, redirect_count, token, url](negotiation_response&& response, std::exception_ptr exception)
+            [callback, weak_connection, redirect_count, token, url, started](negotiation_response&& response, std::exception_ptr exception)
```
Comment (BrennanConroy) [inline src/signalrclient/connection_impl.cpp:215]:
```suggestion
            [callback, weak_connection, redirect_count, token, url, transport_started](negotiation_response&& response, std::exception_ptr exception)
```

File: src/signalrclient/connection_impl.cpp [hunk 5]
```diff
@@ -232,58 +288,7 @@ namespace signalr
...
-                                std::string("internal error - transition from an unexpected state. expected state: connecting, actual state: ")
-                                .append(translate_connection_state(connection->get_connection_state())));
-
-                            assert(false);
-                        }
-
-                        connection->m_start_completed_event.cancel();
-                        callback(nullptr);
-                    });
+                connection->start_transport(url, started);
```
Comment (BrennanConroy) [inline src/signalrclient/connection_impl.cpp:291]:
```suggestion
                connection->start_transport(url, transport_started);
```

**Comment by BrennanConroy** [comment]
Hang fixed 👍 

Could you add a test or two for this?

**Comment by wegylexy** [comment]
I am not familiar with adding tests, and am busy struggling with another project.

**Review by BrennanConroy — APPROVED** [review]

**Comment by BrennanConroy** [comment]
Thanks @wegylexy

**Pull request merged**